cas-server-core/src/main/java/org/jasig/cas/services/DefaultRegisteredServiceAuthorizationStrategy.java

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /**
+     * Set to enable/authorize this service.
+     * @param ssoEnabled true to enable service
+     */
+    public final void setSsoEnabled(final boolean ssoEnabled) {
+        this.ssoEnabled = ssoEnabled;
+    }
+
+    public boolean isEnabled() {
+        return this.enabled;
+    }
+
+    public boolean isSsoEnabled() {
+        return this.ssoEnabled;
+    }
+
+    /**
+     * Defines the attribute aggregation when checking for required attributes.
+     * Default requires that all attributes be present and match the principal's.
+     * @param requireAllAttributes the require all attributes
+     */
+    public final void setRequireAllAttributes(final boolean requireAllAttributes) {
+        this.requireAllAttributes = requireAllAttributes;
+    }
+
+    public final boolean isRequireAllAttributes() {
+        return this.requireAllAttributes;
+    }
+
+    public Map<String, Set<String>> getRequiredAttributes() {
+        return ImmutableMap.copyOf(this.requiredAttributes);
+    }
+
+    /**
+     * Defines the required attribute names and values that
+     * must be available to the principal before the flow
+     * can proceed to the next step. Every attribute in
+     * the map can be linked to multiple values.
+     *
+     * @param requiredAttributes the required attributes
+     */
+    public final void setRequiredAttributes(final Map<String, Set<String>> requiredAttributes) {
+        this.requiredAttributes = requiredAttributes;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Verify presence of service required attributes.
+     * <ul>
+     *     <li>If no required attributes are specified, authz is granted.</li>
+     *     <li>If ALL required attributes must be present, and the principal contains all and there is
+     *     at least one attribute value that matches the required, authz is granted.</li>
+     *     <li>If ALL required attributes don't have to be present, and there is at least
+     *     one principal attribute present whose value matches the required, authz is granted.</li>
+     *     <li>Otherwise, access is denied</li>
+     * </ul>
+     * Note that comparison of principal/required attributes is case-sensitive. Exact matches are required
+     * for any individual attribute value.
+     */
+    @Override
+    public boolean isServiceAccessAuthorizedForPrincipal(final Map<String, Object> principalAttributes, final Service service) {
+        if (this.requiredAttributes.isEmpty()) {
+            logger.debug("No required attributes are specified");
+            return true;
+        }
+        if (principalAttributes.isEmpty()) {
+            logger.warn("No principal attributes are found to satisfy attribute requirements for [{}]", service.getId());
+            return false;
+        }
+
+        if (principalAttributes.size() < this.requiredAttributes.size()) {
+            logger.warn("The size of the principal attributes that are [{}] does not match requirements, "
```

Comment:
why are these all warn?  If we're doing SSO, there's no way that a service can prevent these redirects and thus this could be a common occurrence?   I think the more important thing would be to make sure these are accurately represented in the audit logs.